File: examples/laravel/app/Http/Controllers/VerificationController.php
<?php

namespace App\Http\Controllers;

use App\Exceptions\MyOtpException;
use App\Http\Requests\SendOtpRequest;
use App\Http\Requests\VerifyOtpRequest;
use App\Services\MyOtpClient;
use App\Support\PhoneSanitizer;
use Illuminate\Http\RedirectResponse;
use Illuminate\Http\Request;
use Illuminate\Http\Response;
use Illuminate\View\View;

class VerificationController extends Controller
{
    public function __construct(private readonly MyOtpClient $myotp)
    {
    }

    public function phoneForm(): View
    {
        return view('verification.index');
    }

    public function sendOtp(SendOtpRequest $request): RedirectResponse|Response|View
    {
        $phone = PhoneSanitizer::sanitize($request->input('phone'));
        $channel = $request->input('channel', 'sms');

        if ($phone === '') {
            return response()->view('verification.index', [
                'error' => 'Phone number is required.',
            ], 400);
        }

        try {
            $result = $this->myotp->generateOtp(phoneNumber: $phone, channel: $channel);
        } catch (MyOtpException $e) {
            $status = $e->status >= 400 ? $e->status : 500;
            return response()->view('verification.index', [
                'error' => $e->getMessage(),
            ], $status);
        }

        // message_id is opaque to the browser (a UUID). Laravel encrypts the
        // session cookie when SESSION_ENCRYPT=true (default in this demo) so
        // the value cannot be read or tampered with client-side.
        $request->session()->put('message_id', $result['message_id']);
        $request->session()->put('phone', $phone);
        $request->session()->put('channel', $channel);

        return redirect()->route('verify.form');
    }

    public function verifyForm(Request $request): View|RedirectResponse
    {
        if (! $request->session()->has('message_id')) {
            return redirect()->route('phone.form');
        }

        return view('verification.verify', [
            'phone' => $request->session()->get('phone'),
            'channel' => $request->session()->get('channel'),
        ]);
    }

    public function verifyOtp(VerifyOtpRequest $request): RedirectResponse|Response|View
    {
        if (! $request->session()->has('message_id')) {
            return redirect()->route('phone.form');
        }

        $phone = $request->session()->get('phone');
        $channel = $request->session()->get('channel');
        $messageId = $request->session()->get('message_id');
        $otp = trim((string) $request->input('otp'));

        try {
            $result = $this->myotp->verifyOtp(otp: $otp, messageId: $messageId);
        } catch (MyOtpException $e) {
            $status = $e->status >= 400 ? $e->status : 500;
            return response()->view('verification.verify', [
                'phone' => $phone,
                'channel' => $channel,
                'error' => $e->getMessage(),
            ], $status);
        }

        if (($result['status'] ?? null) !== 'success') {
            return response()->view('verification.verify', [
                'phone' => $phone,
                'channel' => $channel,
                'error' => $result['message'] ?? 'Verification failed.',
            ], 400);
        }

        $request->session()->forget(['message_id', 'channel']);
        $request->session()->put('verified_phone', $phone);
        $request->session()->forget('phone');

        return redirect()->route('verify.success');
    }

    public function success(Request $request): View|RedirectResponse
    {
        $phone = $request->session()->pull('verified_phone');
        if (! $phone) {
            return redirect()->route('phone.form');
        }

        return view('verification.success', ['phone' => $phone]);
    }
}

File: examples/rails/app/controllers/verification_controller.rb
class VerificationController < ApplicationController
  VALID_CHANNELS = %w[sms whatsapp telegram].freeze

  # GET /
  def index
  end

  # POST /send
  def send_otp
    phone = PhoneSanitizer.call(params[:phone])
    channel = VALID_CHANNELS.include?(params[:channel]) ? params[:channel] : "sms"

    if phone.blank?
      @error = "Phone number is required."
      render :index, status: :bad_request
      return
    end

    result = MyotpClient.generate_otp(phone_number: phone, channel: channel)

    # message_id is opaque to the browser (a UUID). Rails signs and encrypts
    # the session cookie with secret_key_base, so it cannot be tampered with.
    session[:message_id] = result.fetch("message_id")
    session[:phone] = phone
    session[:channel] = channel
    redirect_to verify_form_path
  rescue MyotpClient::Error => e
    @error = e.message
    render :index, status: (e.status >= 400 ? e.status : :internal_server_error)
  end

  # GET /verify
  def verify_form
    return redirect_to root_path if session[:message_id].blank?

    @phone = session[:phone]
    @channel = session[:channel]
  end

  # POST /verify
  def verify
    return redirect_to root_path if session[:message_id].blank?

    @phone = session[:phone]
    @channel = session[:channel]
    otp = params[:otp].to_s.strip

    if otp.empty?
      @error = "Enter the code you received."
      render :verify_form, status: :bad_request
      return
    end

    result = MyotpClient.verify_otp(otp: otp, message_id: session[:message_id])

    if result["status"] != "success"
      @error = result["message"] || "Verification failed."
      render :verify_form, status: :bad_request
      return
    end

    verified_phone = session.delete(:phone)
    session.delete(:message_id)
    session.delete(:channel)
    session[:verified_phone] = verified_phone
    redirect_to success_path
  rescue MyotpClient::Error => e
    @error = e.message
    render :verify_form, status: (e.status >= 400 ? e.status : :internal_server_error)
  end

  # GET /success
  def success
    @phone = session.delete(:verified_phone)
    redirect_to(root_path) and return if @phone.blank?
  end
end
